### DISCUSSION TITLE
How to nest Steps and Parallel with Hera Runner?

### DISCUSSION THREAD

**Discussion by bieniekmateusz** [argoproj-labs/hera#1139] [Q&A] [answered]
Hi, I have a workflow that splits into smaller workflows multiple times. 
I see that Steps used as a template is one way to breakdown the workflow. Is there any way to tackle nesting? e.g.

```
with Steps(...):
    with Parallel(...):
        with Steps(name="parallelA"):
            ...
        with Steps(name="parallelB"):
            ...
```

**Comment by flaviuvadan** [comment, marked as answer]
I think https://hera.readthedocs.io/en/stable/examples/workflows/steps/callable_steps/ might help! Argo supports nested _invocations_ of templates but not _definitions_. In the above we have nested definitions of steps whereas in the example all the definitions are defined "at the top level" and invocations are performed in a nested manner

**Reply by flaviuvadan** [reply]
Something like 
```python
@script()
def foo(...): ...

with Workflow(...) as w
    with Steps(...).parallel() as a:
        foo(...)
        foo(...)
    
    with Steps(...).parallel() as b:
        foo(...)
        foo(...)

    with Steps(...).parallel() as main:
        a()
        b()
```

And ofc you can set input parameters to go from main to a and b
